### ISSUE TITLE
Investigate why the Python bindings have memory failures on Mac with the MEMORY_CONSISTENCY check

### ISSUE THREAD

**Issue by winterz** [libical/libical#744]
see https://github.com/libical/libical/issues/738
> The `MEMORY_CONSISTENCY` is also something interesting to me. Maybe python does some magic with the strings and copying them from one (python) object to another can garbage the logic behind this option. I do not know python, this is only a pure guess from my side. I do not know a gobject-introspection tag to note that "this memory can be copied and freed only with this and this function". The non-basic types (like structures) can be registered a `GBoxed` types, which do have these copy & free functions defined, but the string (`char *`) is a basic type and trying to register it as a boxed type might break other things.

I'll create a new issue -- maybe we can find a Mac and/or Python expert to help out.  I don't know much about Python either.  Could be that we need to use the official Python for Mac rather than the python provided by homebrew.  I really don't know.

**Comment by minichma** [comment]
I really don't know much about GLib, Mac and Python but looking at the output in #738 (`pointer being freed was not allocated`) it seems that the inconsistency is that Python/GLib is using `free` or `g_free()` rather than `icalmemory_free_buffer` for freeing memory that has been allocated inside libical using `icalmemory_new_buffer()`. I can't run the tests here right now but I'd assume that in the line `assert value_from_string_clone.as_ical_string() == string` we generate a string that's immediately `free`'ed by the runtime. It seems that this happens using `g_free()`, which causes the problem.

But then again the question would be, why this would work on other platforms and only fail on mac, so my considerations here might be wrong altogether.

Following solutions come to my mind:
* Force GLib to use `icalmemory_free_buffer` for freeing memory that has been transferred from inside libical. This would probably be the cleanest approach and should be quite straight forward. But maybe I'm wrong in my analysis and we already do this anyways. Not sure.
* Deactivate use of test-malloc et al inside the GLib libs even if `MEMORY_CONSISTENCY` is enabled. The `MEMORY_CONSISTENCY` switch is mainly there to ensure that malloc/free are used in a consistent manner *within*  libical core libs, so they can be replaced on platforms like ARM Cortex. Within the GLib libs they will never be replaced, so the check would be nice but is certainly less relevant. Deactivating it should not be a problem.
* What has already been implemented: Disallow `MEMORY_CONSISTENCY` altogether when GLib tests are enabled (basically according to 965d5913d5156568a8b29bc3808c7e4be351a3c9). This has the downside that tests for memory consistency would not run at all, also not inside `regression.c`, etc. if GLib tests are enabled, which seems to be excessive.

**Comment by minichma** [comment]
Managed to run the full test suite, including glib-related tests on Ubuntu. If MEMORY_CONSISTENCY is enabled, all that segfault errors appear.

```
Executing command: /usr/bin/ctest -j14 -C Debug -T test --output-on-failure -R ^libical-glib-value$
[ctest]    Site: 15c50eb470b9
[ctest]    Build name: Linux-c++
[ctest] Test project /home/user/libical/build
[ctest]     Start 62: libical-glib-value
[ctest] 1/1 Test #62: libical-glib-value ...............Subprocess aborted***Exception:   0.05 sec
[ctest] munmap_chunk(): invalid pointer
[ctest] 
[ctest] 
[ctest] 0% tests passed, 1 tests failed out of 1
[ctest] 
[ctest] Total Test time (real) =   0.05 sec
[ctest] 
[ctest] The following tests FAILED:
[ctest] 	 62 - libical-glib-value (Subprocess aborted)
[ctest] Errors while running CTest
[proc] The command: /usr/bin/ctest -j14 -C Debug -T test --output-on-failure -R ^libical-glib-value$ exited with code: 8
[ctest] CTest finished with return code 8
```

So my assumptions in the previous commit seems to be correct that the problem simply is that GLib frees memory via `g_free` or such rather than using `icalmemory_free_buffer`. The reason why the issue disappears when `GOBJECT_INTROSPECTION` isn't defined is that the tests aren't run in this case (see https://github.com/libical/libical/blob/db03cf51fc7544aebab555f4dfbde83afd034ed9/src/test/CMakeLists.txt#L84-L86).

So the problem is not specific to Python and also not to the Mac but to our GLib implementation.

@mcrha FYI

**Comment by mcrha** [comment]
> So the problem is not specific to Python and also not to the Mac but to our GLib implementation.

I'd rather say it's interconnected (except of the Mac, which is truly unrelated here). The base newly allocated strings do not have assigned a free function and they cannot have, or at least I do not know how to do it with the introspection marks. That won't help when using the C API, where the caller frees the string as he/she wants (the `g_free()` is preferred).

I wrote a bit more about it earlier, it's quoted in the description.

If interested, I rewrote the value.py into the C code, [test-value.c](https://github.com/user-attachments/files/17112845/test-value.c.txt), and not all the places with `g_free()` are affected, only the strings returned by `i_cal_value_as_ical_string()` fail, the others are fine.
